notebooks/work-in-progress/ferc1-transform-refactor/ferc1_transform_refactor.py

```diff
@@ -0,0 +1,429 @@
+                df = pd.concat([df, new_col], axis="columns")
+            else:
+                logger.warning(
+                    f"Expected column {col_name} not found in dataframe during table "
+                    "transform."
+                )
+        return df
+
+    return tab_fn
+
+
+def convert_units_col(col: pd.Series, params: UnitConversion) -> pd.Series:
+    """Convert the units of and appropriately rename a column."""
+    new_name = re.sub(pattern=params.pattern, repl=params.repl, string=col.name)
+    # only apply the unit conversion if the column name matched the pattern
+    if new_name == col.name:
+        logger.warning(
+            f"{col.name} did not match the unit rename pattern. Check for typos "
+            "and make sure you're applying the conversion to an appropriate column."
+        )
+    col = (params.multiplier * col) + params.adder
+    col.name = new_name
+    return col
+
+
+convert_units = multicol_transform_fn_factory(convert_units_col, drop=True)
+
+
+def categorize_strings_col(col: pd.Series, params: StringCategories) -> pd.Series:
+    """Impose a controlled vocabulary on freeform string column."""
+    return col.map(params.mapping)
+
+
+categorize_strings = multicol_transform_fn_factory(categorize_strings_col, drop=True)
+
+
+################################################################################
+# TableTransformer classes
+################################################################################
+class AbstractTableTransformer(ABC):
+    """An exmaple abstract base transformer class."""
+
+    table_id: str
+
+    @property
+    def params(self) -> TableTransformParams:
+        """Obtain table transform parameters based on the table ID."""
+        return TableTransformParams.from_id(table_id=self.table_id)
+
+    ################################################################################
+    # Abstract methods that must be defined by subclasses
+    @abstractmethod
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Apply all specified transformations to the appropriate input dataframes."""
+        ...
+
+    ################################################################################
+    # Default method implementations which can be used or overridden by subclasses
+    def rename_columns(self, df: pd.DataFrame, params: RenameColumns) -> pd.DataFrame:
+        """Rename the whole collection of dataframe columns using input params.
+
+        Log if there's any mismatch between the columns in the dataframe, and the
+        columns that have been defined in the mapping for renaming.
+
+        """
+        return df.rename(columns=params.columns)
+
+
+class TestTableTransformerOne(AbstractTableTransformer):
+    """A concrete table transformer class for testing."""
+
+    table_id: str = "test_table_one"
+
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Apply all the Test Transforms."""
+        return (
```

Comment:
I wonder if the `AbstractTableTransformer` could apply the generic transforms so they don't have to be applied each time. Perhaps there could even be some configuration available to the specific Table Transformer classes to blacklist generic transforms if they shouldn't be applied for some reason

Reply:
I'm not sure what you mean here.   The `AbstractTableTransformer` can't be instantiated, because it's an abstract base class. But the default methods it defines are the ones that will be applied unless they're overridden by the subclasses.

Reply:
What I'm thinking about is implementing a method that would apply common transformations, then call the abstract method which will apply table specific transforms. Something like:

```py
...
    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """Apply all generic transformations to the appropriate input dataframes, then apply table specific transforms."""
        return (
            df.pipe(self.rename_columns, self.params.rename_columns)
            .pipe(categorize_strings, self.params.categorize_strings)
            .pipe(convert_units, self.params.convert_units)
            .pipe(self.table_specific_transform)
        )

    @abstract
    def table_specific_transform(self, df: pd.DataFrame) -> pd.DataFrame:
        """Apply all specified transformations to the appropriate input dataframes."""
        ...
```

This way the table specific classes can override the `table_specific_transform` method and not have to worry about applying the same generic transforms for every table. Maybe there will be enough variation in which generic transforms to apply that this isn't really helpful in the real world.

Reply:
Right now I've got 3 layers: the `AbstractTableTransformer` (generic methods that could apply to a variety of tables and datasets) `Ferc1AbstractTableTransformer` (only FERC Form 1 specific methods, which can build upon or override the inherited methods) and concrete classes like `FuelFerc1TableTransformer` which contains any logic that's specific to that particular table.

It turned out that that `FuelFerc1TableTransformer` had to redefine a bunch of stuff because the `fuel_ferc1` table has a bunch of special cases, but I'm hopeful the inherited methods will apply directly to most of the other tables, either on their own or with a call to `super()` such that the inherited bit can be augmented rather than entirely replaced.

Once we have a bigger collection of concrete `TableTransformer` classes I think we should look and see what patterns there are, if any, and how things can be consolidated into the appropriate level in the hierarchy.

Reply:
After a bit more iteration here, it seems like it's appropriate to standardize the way these transformations are applied at the dataset level, but not at the top level `AbstractTableTransformer`. I've set it up now to have `start_transform()`, `main_transform()` and `finish_transform()` methods which are composed to create the top level (no longer abstract) `transform()` method. The `start_transform()` and `finish_transform()` can often be generic to the dataset (e.g. ferc1) and the `main_transform()` is where the table-specific stuff tends to end up.